### PR TITLE
add QK21 keyboard

### PR DESCRIPTION
**Pull request by linlin012**
## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by waffle87** [comment]
A valid `keymaps` directory is missing, which is required.

File: keyboards/onepiece/qk21/logo-JZ.png
Comment (noroadsleft) [inline keyboards/onepiece/qk21/logo-JZ.png:1]: Per the [Keyboard Guidelines](https://docs.qmk.fm/#/hardware_keyboard_guidelines?id=imagehardware-files):

> In an effort to keep the repo size down we’re no longer accepting binary files of any format, with few exceptions. Hosting them elsewhere (such as https://imgur.com/) and linking them in the readme.md is preferred.

Please remove this file (`git rm --cached keyboards/onepiece/qk21/logo-JZ.png`) and host it on an external service.

Comment (linlin012) [reply keyboards/onepiece/qk21/logo-JZ.png:1]: I have updated the document, what should I do to resubmit it? I am a beginner.

File: keyboards/onepiece/qk21/config.h [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+/* disable print */
+//#define NO_PRINT
+#define FORCE_NKRO
+/* disable action features */
+//#define NO_ACTION_LAYER
+//#define NO_ACTION_TAPPING
+//#define NO_ACTION_ONESHOT
+// The pin connected to the data pin of the LEDs
+#define RGB_DI_PIN B7
```
Comment (drashna) [inline keyboards/onepiece/qk21/config.h:22]: This is changing at the end of the month. 
```suggestion
#define RGB_DI_PIN B7
#define WS2812_DI_PIN B7
```
or just:
```
#define WS2812_DI_PIN B7
```
(will fail to compile until develop is merged at end of month)

File: keyboards/onepiece/qk21/logo-JZ.png
Comment (drashna) [reply keyboards/onepiece/qk21/logo-JZ.png:1]: Specifically, any changes to the branch that was used to create the PR will also update the PR, automatically.

File: keyboards/onepiece/qk21/info.json [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    "matrix_pins": {
+        "cols": ["F7", "E6", "C7", "D5"],
+        "rows": ["F0", "F1", "F4", "F5", "F6", "B4"]
+    },
+    "processor": "atmega32u4",
+    "url": "",
+    "usb": {
+        "device_version": "1.0.0",
+        "pid": "0x0050",
+        "vid": "0x1EA7"
```
Comment (drashna) [inline keyboards/onepiece/qk21/info.json:24]: This vendor ID is in use, and should be changed.
https://www.the-sz.com/products/usbid/index.php?v=0x1EA7&p=&n=

Specifically, anything above 0x4000 is unassigned, though the further past this, the better. 

Also, you can check against Vendor ID's used in QMK here: https://yanfali.github.io/qmk_usb_usage/

File: keyboards/onepiece/qk21/qk21.c [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+}, {
+  // LED Index to Flag
+  4, 4, 4, 4,
+  4, 4, 4, 4,
+  4, 4, 4, 4,
+  4, 4, 4, 4,
+  4, 4, 4, 4,
+  4
+}};
```
Comment (drashna) [inline keyboards/onepiece/qk21/qk21.c:45]: This structure should be moved to the info.json file. 

You can run `qmk info -kb onepiece/qk21 -f json` to generate the config, automatically.

File: keyboards/onepiece/qk21/qk21.h
Comment (drashna) [inline keyboards/onepiece/qk21/qk21.h:1]: This file can be completely removed.

File: keyboards/onepiece/qk21/qk21.c [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "qk21.h"
```
Comment (drashna) [inline keyboards/onepiece/qk21/qk21.c:17]:
```suggestion
#include "quantum.h"

```

File: keyboards/onepiece/qk21/readme.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
+# qk21
+
+![qk21](imgur.com image replace me!)
+
+*A short description of the keyboard/project*
+
+* Keyboard Maintainer: [linlin012](https://github.com/linlin012)
+* Hardware Supported: *The PCBs, controllers supported*
+* Hardware Availability: *Links to where you can find this hardware*
```
Comment (drashna) [inline keyboards/onepiece/qk21/readme.md:9]: Should be filled out.

File: keyboards/onepiece/qk21/info.json [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    },
+    "bootmagic": {
+        "enabled": true,
+        "matrix": [0, 0]
+    },
```
Comment (drashna) [inline keyboards/onepiece/qk21/info.json:29]: Is the default values, isn't needed.
```suggestion
```

File: keyboards/onepiece/qk21/info.json [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    "encoder": {
+        "rotary": [
+            { "pin_a": "B1", "pin_b": "B2" }
+        ]
+    }
+    "layouts": {
```
Comment (drashna) [inline keyboards/onepiece/qk21/info.json:35]:
```suggestion
    "community_layouts": ["numpad_6x4"],
    "layouts": {
```

**Review by drashna — CHANGES_REQUESTED** [review]
Is missing a default keymap. That needs to be added.

**Pull request closed without merging**